File: quercus-runtime/src/main/scala/se/oppnadata/portal/quercus/runtime/ServiceModule.scala
package se.oppnadata.portal.quercus.runtime

import scala.collection.JavaConversions._
import com.caucho.quercus.module.AbstractQuercusModule
import com.caucho.quercus.env.Env
import org.osgi.framework.BundleContext
import org.tdr.bootstrap.GlobalReference
import org.osgi.framework.ServiceReference

class ServiceModule extends AbstractQuercusModule {

  println("Initializing OSGi Service module...")
  
  val bundleContext = GlobalReference.get("osgi/bundleContext").asInstanceOf[BundleContext]
  
  def service_ref(env : Env, serviceName : String) : Any = {
  
    // TODO: Use a service tracker here instead?
    // TODO: Write this in Java instead? Does it give anything to have it in Scala here...??
    
    println("Finding reference for service: " + serviceName)
    try {
		var serviceReferences = bundleContext.getAllServiceReferences(null, null)
		for ( serviceRef <- serviceReferences ) { 
			var name = serviceRef.getProperty("serviceName").asInstanceOf[String]
			if ( serviceName.equals(name) ) {
				return this.bundleContext.getService(serviceRef)
			}
		}
	}
	catch {
	  case e : Exception => 
	} 
    return null
  }


}

File: quercus-runtime/src/main/java/se/oppnadata/portal/quercus/runtime/ServiceModule.java
package se.oppnadata.portal.quercus.runtime;

import org.osgi.framework.BundleContext;
import org.osgi.framework.ServiceReference;
import org.tdr.bootstrap.GlobalReference;

import com.caucho.quercus.env.Env;
import com.caucho.quercus.module.AbstractQuercusModule;

public class ServiceModule extends AbstractQuercusModule {

	// TODO: Add commons logging here!!
	private BundleContext bundleContext = (BundleContext) GlobalReference.get("osgi/bundleContext");

	public Object service_ref(Env env, String serviceName) {
		try {
			ServiceReference[] serviceReferences = bundleContext.getAllServiceReferences(null, null);
			for ( ServiceReference serviceRef : serviceReferences ) {
				String name = (String) serviceRef.getProperty("serviceName");
				if ( name != null && serviceName.equals(name) ) {
					return this.bundleContext.getService(serviceRef);
				}
			}
		}
		catch ( Exception e ) {}
		return null;
	}
	
}
